File: web/types/drama.ts
/**
 * Short-drama pipeline entities mirroring `cc-panes-core/src/models/drama.rs`.
 * Shots reference durable media nodes/runs by loose ids; generation state is
 * always read from the media graph.
 */
export interface DramaProject {
  id: string;
  workspaceId: string;
  title: string;
  description: string;
  createdAt: string;
  updatedAt: string;
}

export interface DramaEpisode {
  id: string;
  dramaId: string;
  ordinal: number;
  title: string;
  screenplay: string;
  createdAt: string;
  updatedAt: string;
}

export interface DramaShot {
  id: string;
  episodeId: string;
  ordinal: number;
  title: string;
  dialogue: string;
  prompt: string;
  imageNodeId?: string | null;
  imageRunId?: string | null;
  videoNodeId?: string | null;
  videoRunId?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface CreateDramaProjectRequest {
  workspaceId: string;
  title: string;
  description?: string;
}

export interface UpdateDramaProjectRequest {
  title?: string;
  description?: string;
}

export interface CreateDramaEpisodeRequest {
  dramaId: string;
  title: string;
  ordinal?: number;
  screenplay?: string;
}

export interface UpdateDramaEpisodeRequest {
  title?: string;
  ordinal?: number;
  screenplay?: string;
}

export interface CreateDramaShotRequest {
  episodeId: string;
  ordinal?: number;
  title?: string;
  dialogue?: string;
  prompt?: string;
}

/** Reference fields: empty string clears, undefined keeps the current value. */
export interface UpdateDramaShotRequest {
  ordinal?: number;
  title?: string;
  dialogue?: string;
  prompt?: string;
  imageNodeId?: string;
  imageRunId?: string;
  videoNodeId?: string;
  videoRunId?: string;
}

File: cc-panes-core/src/models/drama.rs
use serde::{Deserialize, Serialize};

/// Short-drama pipeline entities: a drama project groups episodes, an episode
/// owns a screenplay plus ordered shots, and each shot references the durable
/// media nodes/runs generated for it (loose TEXT references — the media graph
/// stays the source of truth for generation state).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DramaProject {
    pub id: String,
    pub workspace_id: String,
    pub title: String,
    pub description: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DramaEpisode {
    pub id: String,
    pub drama_id: String,
    pub ordinal: i64,
    pub title: String,
    pub screenplay: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DramaShot {
    pub id: String,
    pub episode_id: String,
    pub ordinal: i64,
    pub title: String,
    pub dialogue: String,
    pub prompt: String,
    pub image_node_id: Option<String>,
    pub image_run_id: Option<String>,
    pub video_node_id: Option<String>,
    pub video_run_id: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateDramaProjectRequest {
    pub workspace_id: String,
    pub title: String,
    #[serde(default)]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateDramaProjectRequest {
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateDramaEpisodeRequest {
    pub drama_id: String,
    pub title: String,
    #[serde(default)]
    pub ordinal: Option<i64>,
    #[serde(default)]
    pub screenplay: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateDramaEpisodeRequest {
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub ordinal: Option<i64>,
    #[serde(default)]
    pub screenplay: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateDramaShotRequest {
    pub episode_id: String,
    #[serde(default)]
    pub ordinal: Option<i64>,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub dialogue: Option<String>,
    #[serde(default)]
    pub prompt: Option<String>,
}

/// `Some(None)` cannot be expressed through plain JSON, so node/run references
/// are cleared by sending an empty string and set by sending the id.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateDramaShotRequest {
    #[serde(default)]
    pub ordinal: Option<i64>,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub dialogue: Option<String>,
    #[serde(default)]
    pub prompt: Option<String>,
    #[serde(default)]
    pub image_node_id: Option<String>,
    #[serde(default)]
    pub image_run_id: Option<String>,
    #[serde(default)]
    pub video_node_id: Option<String>,
    #[serde(default)]
    pub video_run_id: Option<String>,
}
